### DISCUSSION TITLE
TypeScript 는 JavaScript와 비교했을때 더 빠른가요?

### DISCUSSION THREAD

**Discussion by WooGie911** [wanted-9th-3team/frontend-study#45] [프론트 기술 질문 공간] [answered]
TypeScript 는 JavaScript와 비교했을때 더 빠른가요? 
그렇게 생각한 이유가 무엇인가요?
아는대로 말씀해주세요

**Comment by michoball** [comment]
빠르다는 것의 관점에 따라 다를 것 같습니다.

typescript로 개발할 때 개발에 들어가는 시간이 빠르다고 한다면 맞습니다.  typescript 의 사용으로 javascript로 개발시  뜻하지 않게 발생할 수 있는 많은 오류들을 사전에 컴파일 및 런타임 단계에서 잡을 수 있고 IDE와 연결하여 개발하며 실시간으로 오류코드를 파악할 수 있습니다. 이러한 관점에서는 빠르다고 할 수 있습니다. 

하지만 typescirpt 로 개발한 프로젝트를 실제로 실행할 때는 그냥 js로 개발한 프로젝트보다는 느립니다.  웹 브라우저가 이해하는 언어는 js지 ts가 아닙니다. 따라서 프로젝트를 실행하는 과정에서 ts 파일들을 js로 변환하는 과정이 필요합니다. 
js로 개발한 프로젝트에서는 불필요한 과정이니 느리다고 할 수 있을 것 같습니다. 

하지만 매우 많은 측면에서 ts로 개발하는 것이 프로젝트의 안정성을 높여주고 앞서 얘기한 내용대로 개발하는 시간을 단축할 수 있기 때문에 ts로 개발하는 것이 좋습니다.

**Comment by skyhanull** [comment]
속도실행 관점으로 볼 시 타입스크립트는 자바스크립트의 문법을 기반으로 한 슈퍼셋(superset)이기 때문에 컴파일러를 통해 자바스크립트로 컴파일되어 실행됩니다. 때문에 컴파일을 해주는 시간이 들기때문에 자바스크립트만 사용했을 때보다 속도에서 차이가 납니다.

**Comment by WooGie911** [comment, marked as answer]
타입 스크립트를 왜 쓰느냐와 관련이 있어 보입니다.

기본적으로 타입스크립트는 자바스크립트의 슈퍼셋, 즉 자바스크립트 기본 문법에 타입스크립트의 문법을 추가한 언어입니다. 

자바스크립트는 동적 타입의 인터프리터 언어로 런타임에서 오류를 발견할 수 있습니다. 

이에 반해 타입스크립트는 정적 타입의 컴파일 언어이며 타입스크립트 컴파일러 또는 바벨(Babel)을 통해 자바스크립트 코드로 변환됩니다. 

코드 작성 단계에서 타입을 체크해 오류를 확인할 수 있고 미리 타입을 결정하기 때문에 실행 속도가 매우 빠르다는 장점이 있습니다.

즉, 코드 작성 단계까지 포함해서 본다면 타입스크립트가 빠르다 라고 표현 할 수도 있을 것 같습니다. 

하지만 코드 작성 시 매번 타입을 결정해야 하기 때문에 번거롭고 코드량이 증가하며 컴파일 시간이 오래 걸린다는 단점이 있습니다.

즉, 단순 실행 시간만 본다면 자바스크립트가 더 빠르다고 표현 할 수 있을 것 같습니다.

**Comment by BaekYuri** [comment]
실행속도로 판단한다면, 타입스크립트를 자바스크립트로 컴파일 하는 단계가 추가되기 때문에 타입스크립트보다는 자바스크립트가 빠른 것으로 알고 있습니다.
